flink-runtime/src/main/java/org/apache/flink/runtime/operators/shipping/OutputEmitter.java

```diff
@@ -135,7 +116,7 @@ public OutputEmitter(ShipStrategyType strategy, TypeComparator<T> comparator, Pa
 		case PARTITION_CUSTOM:
 			return customPartition(record.getInstance(), numberOfChannels);
 		case PARTITION_RANGE:
-			return rangePartition(record.getInstance(), numberOfChannels);
+			return rangePartition((Tuple2<Integer, ?>)record.getInstance(), numberOfChannels);
```

Comment:
Yes, the current version of `DataDistribution` was supported by a previous version of the API (even pre-Flink). With the generalized data model, `DataDistribution` should be backed by a `Object[][]` instead of `Key<?>[]` and the `TypeComparator`s should be used to compare the individual key fields.

I think if we change DataDistribution now to operate on `Object[][]`, the follow up task would become much easier. Also if you plan to add a `DataDistribution` later, wouldn't it make more sense to update it now, instead of removing it first and adding an updated version later again?